cms/djangoapps/contentstore/views/item.py

```diff
@@ -631,10 +645,104 @@ def _create_item(request):
+        source_parent (XBlock): A parent of the source XBlock.
+
+    Returns:
+        source_index (int): Index position of the xblock in a parent.
+    """
+    try:
+        source_index = source_parent.children.index(source_usage_key)
+        return source_index
+    except ValueError:
+        return None
+
+
+def _move_item(source_usage_key, target_parent_usage_key, user, source_index=None):
+    """
+    Move an existing xblock as a child of the supplied target_parent_usage_key
+
+    If source_index is provided, xblock would be placed under target_parent_usage_key on that particular position.
+
+    Arguments:
+        source_usage_key (BlockUsageLocator): Locator of source item.
+        dest_usage_key (BlockUsageLocator): Locator of destination item.
+        source_index (int): If provided, insert source item at the provided index location in dest_usage_key item.
+
+    Returns:
+        JsonResponse: Response after the move operation is complete.
+            Response only contains error if an invalid move operation is performed.
+    """
+    # Get the list of all component type XBlocks
+    component_types = sorted(set(name for name, class_ in XBlock.load_classes()) - set(DIRECT_ONLY_CATEGORIES))
+
+    store = modulestore()
+    with store.bulk_operations(source_usage_key.course_key):
+        source_item = store.get_item(source_usage_key)
+        source_parent = source_item.get_parent()
+        target_item = store.get_item(target_parent_usage_key)
+        source_type = source_item.category
+        target_type = target_item.category
+        error = None
+
+        # Store actual/initial index of the source item. This would be sent back with response,
+        # so that with Undo operation, it would easier to move back item to it's original/old index.
+        actual_index = _get_source_index(source_usage_key, source_parent)
+
+        valid_move_type = {
+            'vertical': source_type if source_type in component_types else 'component',
+            'sequential': 'vertical',
+            'chapter': 'sequential',
+        }
+
+        if valid_move_type.get(target_type, '') != source_type:
+            error = _('You can not move {source_type} into {target_type}').format(
+                source_type=source_type,
+                target_type=target_type,
+            )
+        elif source_parent.location == target_item.location:
+            error = _('You can not move an item into the same parent.')
+        elif actual_index is None:
+            error = _('{source_usage_key} not found in {parent_usage_key}').format(
+                source_usage_key=unicode(source_usage_key),
+                parent_usage_key=unicode(source_parent.location)
+            )
+        else:
+            try:
+                source_index = int(source_index) if source_index is not None else None
+            except ValueError:
+                error = _('You must provide source_index as an integer.')
```

Comment:
Nit: this should say: "You must provide target_index as an integer". It would also be good to include the actual value, so that it is possible to see what went wrong in the logs.